server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/ProjectAssociationActions.java

```diff
@@ -0,0 +1,147 @@
+  private static final String PARAM_PROFILE_KEY = "profileKey";
+  private static final String PARAM_PROJECT_KEY = "projectKey";
+  private static final String PARAM_PROJECT_UUID = "projectUuid";
+
+  private final QProfileProjectOperations profileProjectOperations;
+  private final QProfileLookup profileLookup;
+  private final ComponentService componentService;
+  private final Languages languages;
+
+  public ProjectAssociationActions(QProfileProjectOperations profileProjectOperations, QProfileLookup profileLookup, ComponentService componentService, Languages languages) {
+    this.profileProjectOperations = profileProjectOperations;
+    this.profileLookup = profileLookup;
+    this.componentService = componentService;
+    this.languages = languages;
+  }
+
+  void define(WebService.NewController controller) {
+    NewAction addProject = controller.createAction("add_project")
+      .setSince("5.2")
+      .setDescription("Associate a project with a quality profile")
+      .setHandler(new AssociationHandler(profileLookup, componentService) {
+        @Override
+        protected void changeAssociation(String profileKey, String projectUuid, UserSession userSession) {
+          profileProjectOperations.addProject(profileKey, projectUuid, userSession);
+        }
+      });
+    setCommonAttributes(addProject);
+
+    NewAction removeProject = controller.createAction("remove_project")
+      .setSince("5.2")
+      .setDescription("Remove a project's association with a quality profile")
+      .setHandler(new AssociationHandler(profileLookup, componentService) {
+        @Override
+        protected void changeAssociation(String profileKey, String projectUuid, UserSession userSession) {
+          profileProjectOperations.removeProject(profileKey, projectUuid, userSession);
+        }
+      });
+    setCommonAttributes(removeProject);
+  }
+
+  private void setCommonAttributes(NewAction action) {
+    action.setPost(true);
+    action.createParam(PARAM_PROJECT_UUID)
+      .setDescription("A project UUID. Either this parameter, or projectKey must be set");
+    action.createParam(PARAM_PROJECT_KEY)
+      .setDescription("A project key. Either this parameter, or projectUuid must be set");
+    action.createParam(PARAM_PROFILE_KEY)
+      .setDescription("A quality profile key. Either this parameter, or a combination of profileName + language must be set");
+    action.createParam(PARAM_PROFILE_NAME)
+      .setDescription("A quality profile name. If this parameter is set, profileKey must not be set and language must be set to disambiguate");
+    action.createParam(PARAM_LANGUAGE)
+      .setDescription("A quality profile language. If this parameter is set, profileKey must not be set and language must be set to disambiguate")
+      .setPossibleValues(Collections2.transform(Arrays.asList(languages.all()), new Function<Language, String>() {
+        @Override
+        public String apply(Language input) {
```

Comment:
"input" parameter must be marked as @Nullable. For this reason it's better to replace Function by org.sonar.core.util.NonNullInputFunction